### PR TITLE
Support for datalink

### PR DESCRIPTION
**Pull request by andamian**
Reworked ALMA to use the new DataLink service for downloads #1819

### CODE REVIEW

**Comment by bsipocz** [comment]
@andamian - failure is related, could you please add the text file to the repo?

**Comment by andamian** [comment]
@keflavich  @bsipocz - this PR is ready. Integration tests are failing because of ALMA services problems in Europe or so it looks like. I'll investigate in the meantime.

File: docs/alma/alma.rst [hunk 1]
```diff
@@ -285,17 +285,28 @@ cycle 1 data sets tend to be >100 GB!
...
+.. code-block:: python
+
+   >>> link_list = Alma.get_data_info(uids)
+   >>> link_list['content_length'].sum()
+   538298369462
+   >>> len(link_list)
+   >>> 47
 
+By default, ALMA data is delivered as tarball files. However, the content of
+some of these files can be listed and access individually. To get information
```
Comment (keflavich) [inline docs/alma/alma.rst:302]:
```suggestion
some of these files can be listed and accessed individually. To get information
```

**Review by keflavich — APPROVED** [review]
I'm approving, but I  didn't have time to do a deep line-by-line review, I'm just trusting that it works.  THe documentation is clear enough and this seems well-motivated

**Review by bsipocz — CHANGES_REQUESTED** [review]
There are unrelated changes to the submodule got into this, so I leave this quick review in place to make sure we don't yet merge the PR.

**Comment by andamian** [comment]
@bsipocz, @keflavich  - removed the unrelated changes. I think they've got in when I fixed the 'out of date' branch. I always seem to have fun with the submodule :-)
Please hold from merging into master: I'm double checking with ALMA folks that their services are ready for this. I'll get back to you when I receive their confirmation. Thanks!

**Comment by bsipocz** [comment]
I'm also in the middle of a half useless nitpicky review, hopefully will finish it up by then end of the week.

**Comment by keflavich** [comment]
@andamian what's the status on this one right now?

**Review by bsipocz — COMMENTED** [review]
this now needs a rebase for the conflict as well as to pick up CI. I'll leave thorough content review for Adam, only one minor thing from my side.

File: astroquery/alma/core.py [hunk 2]
```diff
@@ -468,18 +483,22 @@ def stage_data(self, uids, expand_tarfiles=False, return_json=False):
         """
         Obtain table of ALMA files
 
+        DEPRECATED: Data is no longer staged. This method is deprecated and
```
[the remark this answers is not in the document]
Comment (andamian) [reply astroquery/alma/core.py:504]: Done

File: astroquery/alma/core.py [hunk 3]
```diff
@@ -341,8 +363,8 @@ def query_async(self, payload, cache=None, public=True, science=True,
             return payload
 
         query = _gen_sql(payload)
-        result = self.query_tap(query, **kwargs)
-        if result:
+        result = self.query_tap(query, maxrec=payload.get('maxrec', None))
+        if result is not None:
```
Comment (keflavich) [inline astroquery/alma/core.py:367]: should the alternative be to raise some sort of exception rather than `return None`, as it presently is?

**Review by keflavich — APPROVED** [review]
green means good, right?  I see a lot of green tests.  I didn't do a careful re-review

**Comment by bsipocz** [comment]
> If I'm not mistaken codecov reports coverage for tests and they don't seem to be quite correct.

yes, ignore codecov for now, we need to figure out what the best way to get it sorted out in this case where a large fraction of the test suite relies on remote data tests.

File: astroquery/alma/core.py (hunk 3 above)
Comment (andamian) [reply astroquery/alma/core.py:367]: True. It should never happen so I've changed it to raise an exception now.

**Comment by andamian** [comment]
I've had this on hold for a week because I've made some changes and could run the int-tests for a while due to a lengthy power outage the affected the `CADC` service. Managed to do it today so it's once again ready.

**Comment by andamian** [comment]
@bsipocz & @keflavich - friendly reminder about PR which I consider ready. Please let me know there's anything left for me to do to make it happen. Thanks

**Comment by keflavich** [comment]
OK I'm going ahead with the merge.

**Pull request merged**